superset/common/query_object.py

```diff
@@ -0,0 +1,56 @@
+# pylint: disable=R
+from typing import Dict, List, Optional, Union
+
+from superset import app
+from superset.utils import core as utils
+
+# TODO: Type Metrics dictionary with TypedDict when it becomes a vanilla python type
+# https://github.com/python/mypy/issues/5288
+Metric = Union[str, Dict]
+
+
+class QueryObject:
+    """
+    The query object's schema matches the interfaces of DB connectors like sqla
+    and druid. The query objects are constructed on the client.
+    """
+    def __init__(
+            self,
+            granularity: str,
+            groupby: List[str],
+            metrics: List[Metric],
+            filters: List[str],
+            time_range: Optional[str] = None,
+            time_shift: Optional[str] = None,
```

Comment:
I don't think we need this though, since we will handle the time_shift in the client.

Reply:
hmm, this should probably be handled on the server like `time_range`. The currently dttm parsing logic is quite complex (see below) and we run the risk of having time-zone related discrepancies if we parse timestamps on the client. 

```
since_dttm, until_dttm = utils.get_since_until(time_range)
time_shift_dttm = utils.parse_human_timedelta(time_shift)
self.from_dttm = None if since_dttm is None else (since_dttm - time_shift_dttm)
self.to_dttm = None if until_dttm is None else (until_dttm - time_shift_dttm)
```
Thoughts?

Reply:
You are right! this is the correct way to do that. Didn't think about the calculation part.